debian/patches/backport-ubuntu-pro-config.patch

```diff
@@ -0,0 +1,2714 @@
+@@ -0,0 +1,627 @@
++# This file is part of cloud-init. See LICENSE file for license information.
++
++"""ubuntu_pro: Configure Ubuntu Pro support services"""
++
++import json
++import logging
++import re
++import sys
++from textwrap import dedent
++from typing import Any
++from urllib.parse import urlparse
++
++from cloudinit import subp, util
++from cloudinit.cloud import Cloud
++from cloudinit.config.schema import (
++    get_schema_doc, validate_cloudconfig_schema)
++from cloudinit.settings import PER_INSTANCE
++
++PRO_URL = "https://ubuntu.com/pro"
++DEPRECATED_KEYS = set(["ubuntu-advantage", "ubuntu_advantage"])
++
++pro_properties = {
++  "type": "object",
++  "additionalProperties": False,
++  "properties": {
++    "commands": {
++      "type": ["object", "array"],  # Array of strings or dict
++      "items": {
++        "oneOf": [
++          {"type": "array", "items": {"type": "string"}},
++          {"type": "string"}
++        ]
++      },
++      'additionalItems': False,  # Reject non-string & non-list
++      'minItems': 1,
++      'minProperties': 1,
++    },
++    "enable": {
++      "type": "array",
++      "items": {
++        "type": "string"
++      },
++      "description": "Optional list of Ubuntu Pro services to enable. Any of: cc-eal, cis, esm-infra, fips, fips-updates, livepatch. By default, a given contract token will automatically enable a number of services, use this list to supplement which services should additionally be enabled. Any service unavailable on a given Ubuntu release or unentitled in a given contract will remain disabled. In Ubuntu Pro instances, if this list is given, then only those services will be enabled, ignoring contract defaults. Passing beta services here will cause an error."
++    },
++    "enable_beta": {
++      "type": "array",
++      "items": {
++        "type": "string"
++      },
++      "description": "Optional list of Ubuntu Pro beta services to enable. By default, a given contract token will automatically enable a number of services, use this list to supplement which services should additionally be enabled. Any service unavailable on a given Ubuntu release or unentitled in a given contract will remain disabled. In Ubuntu Pro instances, if this list is given, then only those services will be enabled, ignoring contract defaults."
++    },
++    "token": {
++      "type": "string",
++      "description": "Contract token obtained from https://ubuntu.com/pro to attach. Required for non-Pro instances."
++    },
++    "features": {
++      "type": "object",
++      "description": "Ubuntu Pro features.",
++      "additionalProperties": False,
++      "properties": {
++        "disable_auto_attach": {
++          "type": "boolean",
++          "description": "Optional boolean for controlling if ua-auto-attach.service (in Ubuntu Pro instances) will be attempted each boot. Default: ``false``",
++          "default": False
++        }
++      }
++    },
++    "config": {
```

Comment:
I don't understand why these are here. Does Pro on Xenial do anything with these? Same question with the `KNOWN_PRO_CONFIG_PROPS`.

Reply:
@TheRealFalcon the Pro client on xenial does support`pro config (set|unset)` as latest Pro client version 27.5 is SRU'd t xenial already so latest pro client cli is already supported there. With this backport PR for cloud-init, Pro customers will be able to now launch cloud-config containing `ubuntu_pro: {config: {k:v, ....}}` and know that [cloud-init can speak `pro config set k=v`](https://github.com/canonical/cloud-init/pull/5600/files#diff-9a65f7ac405261b233a2462001db0bc7d4a06ef84a04edd2ddf1ec2515aa90ecR530) during initial boot.  Original Xenial didn't support this feature, but it it is additive per this backport and a non-breaking change.


Per KNOWN_PRO_CONFIG_PROPS, it is a subset of known supported `pro config` properties on all series and can grow independently as new config options are added to pro client and released everywhere.
 On xenial, pro config show for pro --version 27.5 shows that our subset in cloud-init is well within supported config key names.
```
root@test-x-vm:~# pro config show
http_proxy              None
https_proxy             None
apt_http_proxy          None
apt_https_proxy         None
ua_apt_http_proxy       None
ua_apt_https_proxy      None
global_apt_http_proxy   None
global_apt_https_proxy  None
update_messaging_timer  21600
metering_timer          14400
apt_news                True
apt_news_url            https://motd.ubuntu.com/aptnews.json
```

We can extend those KNOWN_PRO_CONFIG_PROPS keys in tip of main in the future if we want strict schema validation and concise errors in cloud-init, but it is not necessary as our JSON schema allows for extraProperties: True to remain flexible with unknown pro config keys.